### ISSUE TITLE
BUG: BatchNormalization in nested models throws DisconnectedInputError

### ISSUE THREAD

**Issue by berleon** [fchollet/keras#1386]
As mentioned by @Sebubu in #1275 the following code results in a DisconnectedInputError:
He notes removing the BatchNormalization layer fixes the code.
```python
from keras.models import Graph,  Sequential
from keras.layers.core import Dense
from keras.layers.normalization import BatchNormalization
from keras.layers.advanced_activations import PReLU


g = Graph()
g.add_input("input", input_shape=[20])
g.add_node(Dense(10), "dense", "input")
g.add_node(BatchNormalization(),"bn", "dense")
g.add_node(PReLU(),"activ", "bn")
g.add_output("output", "activ")
print "g output " + str(g.output_shape)

g2 = Graph()
g2.add_input("input", input_shape=[10])
g2.add_node(Dense(15), "dense", "input")
g2.add_node(BatchNormalization(),"bn", "dense")
g2.add_node(PReLU(),"activ", "bn")
g2.add_output("output", "activ")

model = Sequential()

model.add(g)
model.add(g2)

model.compile(loss="mse",optimizer="adadelta")
```

The problem here is that in the `BatchNormalization` layer the `update` rules of the `running_mean` and `running_std` members are set in the `build` method. When the input of the batch norm layer changes, the `update` rules are not updated accordingly.

**Comment by Sebubu** [comment]
Nice, you've found another fault.Thank you for your effort.
I've tested it and it seems working.

The more general problem is the call of the method get_input(), which force the graph to cache the output of the previous layers. The cache is not cleared in model.add().

My suggestion:
- Clear layer caches before compiling. Could be a troublesome work.
- Rerun build() method after clearing. This would obviate your fix.

**Comment by berleon** [comment]
I close the issue as # #1387 is merged. If you have a better way to resolve this issue, feel free to reopen it.
